src/Http/WebUtilities/src/FormPipeReader.cs

```diff
@@ -0,0 +1,388 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Internal;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    /// <summary>
+    /// Used to read an 'application/x-www-form-urlencoded' form.
+    /// Internally reads from a PipeReader.
+    /// </summary>
+    public class FormPipeReader
+    {
+        private const int StackAllocThreshold = 128;
+        private const int DefaultValueCountLimit = 1024;
+        private const int DefaultKeyLengthLimit = 1024 * 2;
+        private const int DefaultValueLengthLimit = 1024 * 1024 * 4;
+
+        // Used for UTF8/ASCII (precalculated for fast path)
+        private static ReadOnlySpan<byte> UTF8EqualEncoded => new byte[] { (byte)'=' };
+        private static ReadOnlySpan<byte> UTF8AndEncoded => new byte[] { (byte)'&' };
+
+        // Used for other encodings
+        private ReadOnlyMemory<byte> _otherEqualEncoding;
```

Comment:
nit; Just store these as byte[], it's much more efficient (even though this is the slow path).